### ISSUE TITLE
unimplemented system call 11 (munmap)

### ISSUE THREAD

**Issue by justinc1** [cloudius-systems/osv#658]
Loading helloworld OpemMPI application, I get:
```
syscall(): unimplemented system call 11. Aborting.

[backtrace]
0x000000000022bdcf <abort(char const*, ...)+249>
0x000000000059e26c <syscall+4490>
0x000010000609ff78 <???+101318520>
```

Looking at OSv linux.cc and musl/arch/x86_64/bits/syscall.h, this should be __NR_munmap.
Possible to add it?

**Comment by copumpkin** [comment]
Are they calling `munmap` via something other than libc?

**Comment by justinc1** [comment]
with gdb catch on linux:
```
(gdb) catch syscall 11
Catchpoint 1 (syscall 'munmap' [11])
#0  0x00007ffff7df3687 in munmap () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007ffff7df13a8 in _dl_unload_cache () at dl-cache.c:311
#2  0x00007ffff7ddd96f in dl_main (phdr=<optimized out>, phdr@entry=0x400040, phnum=<optimized out>, phnum@entry=9, user_entry=user_entry@entry=0x7fffffffe448, auxv=<optimized out>) at rtld.c:2310
#3  0x00007ffff7df1565 in _dl_sysdep_start (start_argptr=start_argptr@entry=0x7fffffffe530, dl_main=dl_main@entry=0x7ffff7ddb910 <dl_main>) at ../elf/dl-sysdep.c:249
#4  0x00007ffff7ddecf8 in _dl_start_final (arg=0x7fffffffe530) at rtld.c:332
#5  _dl_start (arg=0x7fffffffe530) at rtld.c:558
#6  0x00007ffff7ddb2d8 in _start () from /lib64/ld-linux-x86-64.so.2
#7  0x0000000000000001 in ?? ()
#8  0x00007fffffffe7ab in ?? ()
#9  0x0000000000000000 in ?? ()
```
So it seems to go via glibc. Right?

**Comment by justinc1** [comment]
I assume code from ```musl/src/mman/munmap.c``` should be used, but somehow it is not.

(And since I got error about no shm_open - that one is in the musl/src/mman/shm_open.c, but somehow it not used? I added syntax error there, and scripts/build didn't even complain)

**Comment by nyh** [comment]
It's easy to implement syscall(NR_munmap) in linux.cc, but looking at your stack trace, I'm afraid your problem is something different: it seems you've actually loaded "/lib64/ld-linux-x86-64.so.2" from Linux, and you shouldn't have - OSv already supplies these symbols, and "ld-linux-x86-64.so.2" should not even be in your OSv image.

Can you try to remove this file from your image?

I wonder why it was even loaded. in core/elf.cc we have a list of libraries which OSv itself already supplies so should not be loaded, and ld-linux-x86-64.so.2 is on that list.

**Comment by justinc1** [comment]
@nyh Sorry I wasn't clear - the above trace was actually run on linux, not on OSv - so yes, linux lib was used. I guess with gdb extension/plugin (that comes with OSv), I could run something like 'gdb ./mpi_hello.so'?

PS: For time being, I just added a do-nothing stub for munmap.

**Comment by nyh** [comment]
On Fri, Oct 2, 2015 at 6:05 PM, justinc1 <notifications@github.com> wrote:

> @nyh <https://github.com/nyh> Sorry I wasn't clear - the above trace was
> actually run on linux, not on OSv
>
- so yes, linux lib was used. I guess with gdb extension/plugin (that comes
> with OSv), I could run something like 'gdb ./mpi_hello.so'?
>
Yes, it's quite easy to get a backtrace of the OSv abort in OSv. This (and
more) is explained in:

https://github.com/cloudius-systems/osv/wiki/Debugging-OSv

PS: For time being, I just added a do-nothing stub for munmap.
>
In linux.cc? Why not just call the real munmap, like all the other system
calls there?

**Comment by justinc1** [comment]
I missed real munmap (didn't really search for it enought, when linker complained 'no symbol'). A trivial patch at https://github.com/justinc1/osv/commit/c8e036bc7bff19415974bd8a1573cb43fa7f03d5 looks OK?
Should be a mmap added to linux.cc too?

**Comment by nyh** [comment]
On Wed, Oct 7, 2015 at 3:20 PM, justinc1 <notifications@github.com> wrote:

> I missed real munmap (didn't really search for it enought, when linker
> complained 'no symbol'). A trivial patch at justinc1@c8e036b
> <https://github.com/justinc1/osv/commit/c8e036bc7bff19415974bd8a1573cb43fa7f03d5>
> looks OK?
>

Reviewed-by: Nadav Har'El <nyh@scylladb.com>

Can you please send the patch to this mailing list, so the committers
(Pekka or Avi) can commit it?

Should be a mmap added to linux.cc too?
>

You can.

**Comment by penberg** [comment]
AFAICT, this is fixed by commits 43536e17f860e0bc0029558dba3992999c09f3ed and 443661745368dd0ea7e024a3f300bf0b5375ffa6. @justinc1, can you please close the issue?
